crates/icrate/examples/metal.rs

```diff
@@ -0,0 +1,365 @@
+
+            // create the pipeline state
+            let pipeline_state = device
+                .newRenderPipelineStateWithDescriptor_error(&pipeline_descriptor)
+                .expect("Failed to create a pipeline state.");
+
+            // configure the metal view delegate
+            unsafe {
+                let object = ProtocolObject::from_ref(self);
+                mtk_view.setDelegate(Some(object));
+            }
+
+            // configure the window
+            unsafe {
+                window.setContentView(Some(&mtk_view));
+                window.center();
+                window.setTitle(ns_string!("metal example"));
+                window.makeKeyAndOrderFront(None);
+            }
+
+            // initialize the delegate state
+            self.command_queue.borrow_mut().replace(command_queue);
+            self.pipeline_state.borrow_mut().replace(pipeline_state);
+            self.window.borrow_mut().replace(window);
+            self.mtk_view.borrow_mut().replace(mtk_view);
+        }
+    }
+
+    // define the delegate methods for the `MTKViewDelegate` protocol
+    unsafe impl MTKViewDelegate for Delegate {
+        #[method(drawInMTKView:)]
+        #[allow(non_snake_case)]
+        unsafe fn drawInMTKView(&self, _view: &MTKView) {
+            let command_queue = self.command_queue.borrow();
+            let Some(command_queue) = command_queue.as_ref() else { return; };
+
+            let mtk_view = self.mtk_view.borrow();
+            let Some(mtk_view) = mtk_view.as_ref() else { return; };
+
+            let pipeline_state = self.pipeline_state.borrow();
+            let Some(pipeline_state) = pipeline_state.as_ref() else { return; };
+
+            // FIXME: icrate `MTKView` doesn't have a generated binding for `currentDrawable` yet
+            // (because it needs a definition of `CAMetalDrawable`, which we don't support yet) so
+            // we have to use a raw `msg_send_id` call here instead.
+            let current_drawable: Option<Id<ProtocolObject<dyn MTLDrawable>>> =
+                msg_send_id![mtk_view, currentDrawable];
+
+            // prepare for drawing
+            let Some(current_drawable) = current_drawable else { return; };
+            let Some(command_buffer) = command_queue.commandBuffer() else { return; };
+            let Some(pass_descriptor) = (unsafe { mtk_view.currentRenderPassDescriptor() }) else { return; };
+            let Some(encoder) = command_buffer.renderCommandEncoderWithDescriptor(&pass_descriptor) else { return; };
+
+            // compute the scene properties
+            let mut scene_properties_data = SceneProperties {
+                time: unsafe { self.start_date.timeIntervalSinceNow() } as f32,
+            };
+            // write the scene properties to the vertex shader argument buffer at index 0
+            let scene_properties_bytes =
+                unsafe { NonNull::new_unchecked(&mut scene_properties_data) };
```

Comment:
Hmm, I think these are copied by `setVertexBytes_length_atIndex`. So probably would be better to use `&` and the [`From<&T>` impl on `NonNull`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#impl-From%3C%26T%3E-for-NonNull%3CT%3E) here instead.

```suggestion
            let scene_properties_data = SceneProperties {
                time: unsafe { self.start_date.timeIntervalSinceNow() } as f32,
            };
            // write the scene properties to the vertex shader argument buffer at index 0
            let scene_properties_bytes = NonNull::from(&scene_properties_data);
```

Actually, maybe we should derive `Copy` for `SceneProperties`, just to make that copying very clear?

Same goes for `VertexInput` stuff below.